### PR TITLE
Fix go get response if only app URL is custom in configuration

### LINKED ISSUE #2633: godoc not showing documentation for gitea repos
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes
  - [ ] No
  - [ ] Not relevant

## Description

https://try.gitea.io/Howl/gorepo is a valid go package. However, godoc.org does not recognise it as such - and renders a 404 https://godoc.org/try.gitea.io/Howl/gorepo

Here is proof that godoc.org still works with other package urls:

* https://godoc.org/robpike.io/ivy
* https://godoc.org/try.gogs.io/gogs/gogs

### PR DESCRIPTION
**Pull request by lafriks**
Fixes #2633

### CODE REVIEW

File: modules/context/repo.go [hunk 1]
```diff
@@ -136,7 +137,11 @@ func RetrieveBaseRepo(ctx *Context, repo *models.Repository) {
 
 // ComposeGoGetImport returns go-get-import meta content.
 func ComposeGoGetImport(owner, repo string) string {
-	return path.Join(setting.Domain, setting.AppSubURL, owner, repo)
+	baseURL, err := url.Parse(setting.AppURL)
+	if err != nil {
+		path.Join(setting.Domain, setting.AppSubURL, owner, repo)
```
Comment (lunny) [inline modules/context/repo.go]: need return?

**Comment by lafriks** [comment]
@lunny fixed

**Comment by ethantkoenig** [comment]
LGTM

**Comment by sapk** [comment]
LGTM

For futur work, couldn't we set setting.Domain at appURL.Hostname globally (if domain is not set in ini) ? If it's ok I could do it.

**Comment by lafriks** [comment]
@ethantkoenig @sapk rewritten as requested and setting.Domain should now always match domain in AppURL except when AppURL has IP adress not domain in host part of URL.

File: modules/setting/setting.go [hunk 2]
```diff
@@ -664,6 +665,12 @@ func NewContext() {
 	// This value is empty if site does not have sub-url.
 	AppSubURL = strings.TrimSuffix(url.Path, "/")
 	AppSubURLDepth = strings.Count(AppSubURL, "/")
+	// Check if Domain differs from AppURL domain than update it to AppURL's domain
+	// TODO: Can be replaced with url.Hostname() when minimal GoLang version is 1.8
+	urlHostname := strings.SplitN(url.Host, ":", 2)[0]
+	if urlHostname != Domain && net.ParseIP(urlHostname) == nil {
+		Domain = urlHostname
```
Comment (sapk) [inline modules/setting/setting.go:672]: Domain as it used should contain port at least for the use maid in : https://github.com/go-gitea/gitea/blob/2ef8b8bcbeeb7efabb6cfd247ad6227b0e2e21cd/modules/context/repo.go#L138 
`path.Join(setting.Domain, setting.AppSubURL, owner, repo)`

Comment (sapk) [reply modules/setting/setting.go:672]: It would be better to update the ComposeGoGetImport to add HTTP_PORT

Comment (lafriks) [reply modules/setting/setting.go:672]: @sapk no it should not contain port as it generates `go` `import` part of repository and go imports can not contain port

Comment (sapk) [reply modules/setting/setting.go:672]: Sorry mislead HTTP_PORT is not necessary the same as the public port ... but we still need to add the url port to the go-import.

Comment (sapk) [reply modules/setting/setting.go:672]: ok sorry that kind of strange how it works if the host is not on defaut port ?

Comment (sapk) [reply modules/setting/setting.go:672]: oh sorry I understand that the repo url after package that contain the port.

**Review by sapk — APPROVED** [review]

File: modules/setting/setting.go (hunk 2 above)
Comment (lafriks) [reply modules/setting/setting.go:672]: Yes git url is generated in other place. This is just import part

**Comment by sapk** [comment]
New version LGTM.

**Pull request merged**